### ISSUE TITLE
HDTileBasedDataStore performance - reading waynodes is slow

### ISSUE THREAD

**Issue by mapsforge** [mapsforge/mapsforge#194]
_From [Dennis.K...@gmail.com](https://code.google.com/u/115287360692073326709/) on October 20, 2011 17:26:45_

What steps will reproduce the problem? 1. Produce the map from relatively large OSM data file (1GB xml) using type=hd.

Actual behavior:
The execution of complete() method takes significant amount of time and does not utilize computer resources. What version of the product are you using? On what operating system? 0.2.4, Windows 7 x64 Please provide any additional information below. Here is the time distribution of processing 1GB xml file:
1. Reading data: 10 min
2. complete() function: 5 hours 
3. Writing file: 6 hours

The most of the time (99&#37;) in the complete() function is spent on finding way nodes by ID in the IndexedObjectStore structure. 
Call stack:
- org.mapsforge.preprocessing.map.osmosis.HDTileBasedDataStore.complete
- org.mapsforge.preprocessing.map.osmosis.TileData$TDWay.fromWay
- org.mapsforge.preprocessing.map.osmosis.HDTileBasedDataStore.getEntity
- org.openstreetmap.osmosis.core.store.IndexedObjectStoreReader.get

During this operation the machine resources (CPU, memory, IO) are not utilized at all (except for small IO reads).

There is a number of way to increase the performance of the complete() function:
1. Figure out why IndexedObjectStore is so slow finding the objects by ID.
2. Use the producer-consumer model and perform way processing in parallel.

_Original issue: http://code.google.com/p/mapsforge/issues/detail?id=194_

**Comment by devemux86** [comment]
Mapsforge has changed a lot since then, can create new issue if still valid.
